src/languages/en.ts

```diff
@@ -4645,4 +4703,6 @@ export default {
         notAllowedMessageHyperLinked: ' limited access',
         notAllowedMessageEnd: ' copilot',
     },
-} satisfies TranslationBase;
+};
+
+export default translations satisfies TranslationBase<typeof translations>;
```

Comment:
Okay now I understand, you moved the export under here for a reason hahah

If you try this:
```ts
const en = {
    test: 'lol',
    common: {
        yes: 'Yes',
        x: ({x}: {x: string}) => `should work ${x}`,
        j: ({test}: {test: string}, {test2}: {test2: string}) => `should fail, two argument not allowed ${test} ${test2}`,
        i: (num: number) => `should fail, argument is not an object ${num} ${ts}`,
        l: (num: Set<number>) => `should fail, argument is not an object ${num} ${ts}`,
        k: (x) => `should fail, argument is not an object ${x}`,
    },
} satisfies TranslationBase2<typeof en>;
```

You get an error:
`'en' implicitly has type 'any' because it does not have a type annotation and is referenced directly or indirectly in its own initializer.`

Reply:
Yeah 
I tried different ways
But my current version is the only one that works 😅

Reply:
Great idea
Thanks !
But for better understanding
I added many conditions
To make ts error text relevant to the issue itself

Reply:
I tried to simplify it, would that work? Do you see any bugs?

```ts
/**
 * Retrieves the first argument of a function
 */
type FirstArgument<TFunction> = TFunction extends (arg: infer A, ...args: any[]) => any ? A : never;

/**
 * Translation value can be a string or a function that returns a string
 */
type TranslationLeafValue<TStringOrFunction> = TStringOrFunction extends string
    ? string
    : (
          arg: FirstArgument<TStringOrFunction> extends Record<string, unknown> ? FirstArgument<TStringOrFunction> : Record<string, unknown>,
          ...noOtherArguments: unknown[]
      ) => string | PluralForm;

/**
 * Translation object is a recursive object that can contain other objects or string/function values
 */
type TranslationDeepObject<TTranslations> = {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    [Path in keyof TTranslations]: TTranslations[Path] extends string | ((...args: any[]) => any) ? TranslationLeafValue<TTranslations[Path]> : TranslationDeepObject<TTranslations[Path]>;
};

const en2 = {
    test: 'lol',
    common: {
        yes: 'Yes',
        x: ({x}: {x: string}) => `should work ${x}`,
        plural: () => ({
            zero: 'No expenses',
            one: 'One expense',
            other: (count: number) => `${count} expenses`,
        }),
        // j: ({test}: {test: string}, {test2}: {test2: string}) => `should fail, two argument not allowed ${test} ${test2}`,
        // i: (num: number) => `should fail, argument is not an object ${num}`,
        // l: (sett: Set<number>) => `should fail, argument is not an object ${sett.has(1)}`,
        k: (x) => `should fail, argument is not an object ${x}`,
    },
};

const x = en2 satisfies TranslationDeepObject<typeof en2>;
```

Reply:
But it looks better anyway 😅

Reply:
@ZhenjaHorbach Are you sure? It errors for me 😲 

![Screenshot 2024-09-11 at 19 34 09](https://github.com/user-attachments/assets/352282e1-87fa-468f-9b8d-c381151a0701)

Reply:
Oh yeah 
Restart ts server 
And now I see the issue 😅
I continue the test !

Reply:
I found one problem, it doesn't work when a default value is passed 🤔 
```ts
const en2 = {
    test: 'lol',
    common: {
           deleteExpense: ({count}?: {count: number} = {count: 1}) => `Delete ${count}`,
    },
};

const x = en2 satisfies TranslationDeepObject<typeof en2>;
```

Reply:
It works interesting
If this is an optional parameter
Then everything is fine

<img width="790" alt="Снимок экрана 2024-09-11 в 19 42 27" src="https://github.com/user-attachments/assets/ea4c97f1-3f64-4e94-8d89-3570cf3c9932">

Reply:
Yep, found a fix!
```ts
/**
 * Retrieves the first argument of a function
 */
type FirstArgument<TFunction> = TFunction extends (arg: infer A, ...args: any[]) => any ? A : never;

/**
 * Translation value can be a string or a function that returns a string
 */
type TranslationLeafValue<TStringOrFunction> = TStringOrFunction extends string
    ? string
    : (
          arg: FirstArgument<TStringOrFunction> extends Record<string, unknown> | undefined ? FirstArgument<TStringOrFunction> : Record<string, unknown>,
          ...noOtherArguments: unknown[]
      ) => string | PluralForm;

/**
 * Translation object is a recursive object that can contain other objects or string/function values
 */
type TranslationDeepObject<TTranslations> = {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    [Path in keyof TTranslations]: TTranslations[Path] extends string | ((...args: any[]) => any) ? TranslationLeafValue<TTranslations[Path]> : TranslationDeepObject<TTranslations[Path]>;
};
```

Reply:
```
const translations = {
    test: 'lol',
    common: {
        yes: 'Yes',
        k: ({value}: {value: string}) => `should fail, argument is not an object ${value}`,
        deleteExpense: (count?: string) => `Delete ${count}`,
    },
};

export default translations satisfies TranslationBase<typeof translations>;
```

If we pass an optional parameter
There will be no ts error here

Reply:
Actually works great!

Reply:
U sure? 😅 
![Screenshot 2024-09-11 at 19 52 49](https://github.com/user-attachments/assets/7f4f583c-853f-4d77-a8b9-9034c42d890c)

Reply:
I  removed the comment 😅
VS code is not feeling well today

Reply:
typecheck also passed successfully !

Reply:
Merged 
@blazejkustra 
Thanks for helping !

@shubham1206agra 
You can test it too 😃
And if everything is good
Let's fill out the checklist !